### PR TITLE
Build SYCL kernels for ATen XPU ops on Native Windows

### PR DESCRIPTION
**Pull request by min-jean-cho**
Moved to https://github.com/pytorch/pytorch/pull/127390

-------
As proposed in https://github.com/pytorch/pytorch/issues/126719, we are enabling PyTorch XPU on Native Windows on Intel GPU. 

This PR  enables XPU build on Windows as the first step of #126719:

- Enable `USE_XPU` build on Windows using MSVC as host compiler. The use of MSVC as host compiler seamlessly aligns with the existing PyTorch build on Windows.
- Build oneDNN GPU library on Windows.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

File: c10/xpu/XPUFunctions.cpp
Comment (guangyey) [inline c10/xpu/XPUFunctions.cpp:61]: FYI: please rebase code if [126752](https://github.com/pytorch/pytorch/pull/126752) is landed.

File: c10/util/Float8_fnuz_cvt.h
Comment (min-jean-cho) [inline c10/util/Float8_fnuz_cvt.h:37]: @ratnampa please upstream fix to this, and I'll rebase

Comment (gujinghui) [reply c10/util/Float8_fnuz_cvt.h:37]: This line is totally same with line #43, why we do not reuse it?

Comment (min-jean-cho) [reply c10/util/Float8_fnuz_cvt.h:37]: Reason is, we need to use this path for SYCL device code before `_MSC_VER` path for host code.

File: c10/xpu/XPUFunctions.cpp
Comment (guangyey) [reply c10/xpu/XPUFunctions.cpp:61]: hi @min-jean-cho could you help merge https://github.com/pytorch/pytorch/pull/126752 to this PR?

File: c10/util/Float8_fnuz_cvt.h
Comment (gujinghui) [reply c10/util/Float8_fnuz_cvt.h:37]: Better as below?

```suggestion
#elif defined(_MSC_VER) and not defined(__SYCL_DEVICE_ONLY__)
    unsigned long nonsign_bsr;
    _BitScanReverse(&nonsign_bsr, (unsigned long)mantissa);
    uint32_t renorm_shift = (uint32_t)nonsign_bsr ^ 31;
#else
    uint32_t renorm_shift = __builtin_clz(mantissa);
```

Comment (min-jean-cho) [reply c10/util/Float8_fnuz_cvt.h:42]: Yes, thanks.

**Comment by guangyey** [comment]
@min-jean-cho this PR https://github.com/pytorch/pytorch/pull/126838 needs to be merged here as well.

File: cmake/Modules/FindSYCLToolkit.cmake
Comment (dvrogozh) [inline cmake/Modules/FindSYCLToolkit.cmake:54]: which issue and which workaround? please, update comment to give a hint to other developers on what is being awaited. If there is a public issue filed somewhere at github, I suggest to add a link.

File: c10/xpu/XPUMacros.h
Comment (dvrogozh) [inline c10/xpu/XPUMacros.h:17]: Seems to be a typo:
```suggestion
#define C10_XPU_EXPORT
#define C10_XPU_IMPORT
```

**Review by jgong5 — CHANGES_REQUESTED** [review]
Please fix CI failures which seem real.

**Comment by dvrogozh** [comment]
That draft PR currently... But maybe this PR needs to wait for #126410 to be merged first, then rebased to make sure there is no regressions in XPU build for Linux.

**Comment by jgong5** [comment]
> That draft PR currently... But maybe this PR needs to wait for #126410 to be merged first, then rebased to make sure there is no regressions in XPU build for Linux.

We shouldn't break existing CIs at least.

**Comment by dvrogozh** [comment]
> We shouldn't break existing CIs at least.

Definitely.

**Comment by chuanqi129** [comment]
> That draft PR currently... But maybe this PR needs to wait for #126410 to be merged first, then rebased to make sure there is no regressions in XPU build for Linux.

No need waiting for this PR merge, can trigger current XPU build and test by adding `ciflow/xpu` label

File: c10/xpu/XPUMacros.h
Comment (min-jean-cho) [reply c10/xpu/XPUMacros.h:17]: Thanks for the catch.

File: c10/xpu/XPUFunctions.cpp
Comment (min-jean-cho) [reply c10/xpu/XPUFunctions.cpp]: Done, thanks.

**Comment by min-jean-cho** [comment]
> @min-jean-cho this PR https://github.com/pytorch/pytorch/pull/126838 needs to be merged here as well.

@guangyey Done, thanks.

File: cmake/Modules/FindSYCLToolkit.cmake
Comment (min-jean-cho) [reply cmake/Modules/FindSYCLToolkit.cmake]: Thanks, updated comment after offline sync with compiler team.

**Comment by min-jean-cho** [comment]
> Please fix CI failures which seem real.

@jgong5 CI passes, thanks. 

> can trigger current XPU build and test by adding ciflow/xpu label

@chuanqi129 XPU CI passes, thanks.

File: c10/xpu/CMakeLists.txt
Comment (guangyey) [inline c10/xpu/CMakeLists.txt:56]: need a blank line.

File: c10/xpu/XPUFunctions.cpp
Comment (guangyey) [inline c10/xpu/XPUFunctions.cpp:9]:
```suggestion
```
I guess these files are useless.

File: c10/xpu/impl/xpu_cmake_macros.h.in
Comment (guangyey) [inline c10/xpu/impl/xpu_cmake_macros.h.in:6]: need a blank line.

**Review by guangyey — APPROVED** [review]
Others looks good to me.

**Comment by min-jean-cho** [comment]
@jgong5 CI passes. Could you please help review and approve if looks good? Thanks.

**Review by dvrogozh — APPROVED** [review]
looks good

File: cmake/Modules/FindSYCLToolkit.cmake
Comment (jgong5) [inline cmake/Modules/FindSYCLToolkit.cmake:55]: Is it good to assume the version of sycl to `7` here? What's the problem of original implementation?

File: torch/csrc/xpu/Module.cpp
Comment (jgong5) [inline torch/csrc/xpu/Module.cpp:20]: How do we address the bad fork on Windows?

File: aten/src/ATen/xpu/detail/XPUHooks.cpp
Comment (jgong5) [inline aten/src/ATen/xpu/detail/XPUHooks.cpp:50]: How do we plan to address the default context on Windows? Is it in the next step? If so, suggest to add a TODO here.

File: torch/csrc/xpu/Module.cpp
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp:20]: Align to other backends, it is not supported on Windows.

File: aten/src/ATen/xpu/detail/XPUHooks.cpp
Comment (gujinghui) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:50]: It will be not supported in a long time. We will take it as a feature enabling in future when it is available in SYCL on Windows.

Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:50]: I think it is our next step. @min-jean-cho  Could you help to add `TODO`  on line 47 and line 28, like:
`TODO: We will switch to the default context on Windows to support this feature`

File: c10/util/Float8_fnuz_cvt.h
Comment (EikanWang) [inline c10/util/Float8_fnuz_cvt.h:36]: Does it mean the renorm_shift will be `__builtin_clz(mantissa);` for Intel GPU on Windows and why?

File: cmake/Modules/FindMKLDNN.cmake
Comment (EikanWang) [inline cmake/Modules/FindMKLDNN.cmake:31]: Please check with oneDNN team what the `DEFAULT` means. Regarding Linux, it means the device compiler and host compiler are both g++. For Windows, we need to clearly understand the behavior in terms of separate compilation: MSVC for host code compilation, and ICX for device code compilation.

File: cmake/Modules/FindMKLDNN.cmake
Comment (EikanWang) [inline cmake/Modules/FindMKLDNN.cmake:41]: Please do **NOT** use any `DPCPP` in the code.

File: cmake/Modules/FindSYCLToolkit.cmake
Comment (EikanWang) [reply cmake/Modules/FindSYCLToolkit.cmake:55]: It does not make sense to hard code a particular version here.

Comment (min-jean-cho) [reply cmake/Modules/FindSYCLToolkit.cmake:55]: The current version of SYCL runtime from oneAPI is `7`, so I've hardcoded to `7`. 

Problem with the original implementation is that there's no `sycl.dll` on Windows. Only `sycl7.dll` with version suffix. `sycl7.dll` on Windows corresponds to `libsycl.so` on Linux.
In the future, if compiler team adds support to `sycl.dll` on Windows, we can drop this. Otherwise, we need to update `sycl_lib_version` to `8` and so on when SYCL runtime version uplifts.

File: torch/csrc/xpu/Module.cpp
Comment (min-jean-cho) [reply torch/csrc/xpu/Module.cpp:20]: Yes, aligned with other backends.

**Review by EikanWang — CHANGES_REQUESTED** [review]
Please add some descriptions to your PR.

File: c10/util/Float8_fnuz_cvt.h
Comment (min-jean-cho) [reply c10/util/Float8_fnuz_cvt.h:36]: Yes, it means device code will be in `__builtin_clz` while host code will be in `_BitScanReverse`. Reason is, `_BitScanReverse` intrinsic is not supported with SYCL/DPC++ compiler.

File: cmake/Modules/FindMKLDNN.cmake
Comment (min-jean-cho) [reply cmake/Modules/FindMKLDNN.cmake:31]: Currently, `DEFAULT` is the only supported option for `ONEDNN_DPCPP_HOST_COMPILER` on Windows. I will check with oneDNN team and update.

Comment (EikanWang) [reply cmake/Modules/FindMKLDNN.cmake:31]: We may need to submit a requirement to oneDNN team to support separation compilation for Widnows. Please check with @ZhiweiYan-96 .

File: c10/util/Float8_fnuz_cvt.h
Comment (EikanWang) [reply c10/util/Float8_fnuz_cvt.h:36]: The utility function is marked with `C10_HOST_DEVICE`. For `HOST`, we should build it by MSVC. For `DEVICE`, it is only for CUDA. According to your description, Does it mean we build the function by `ICX` and take it as a device kernel? If it is XPU device kernel, we need to refine the function.

File: cmake/Modules/FindMKLDNN.cmake
Comment (min-jean-cho) [reply cmake/Modules/FindMKLDNN.cmake]: Thanks, updated to `SYCL_CXX_DRIVER`.

File: cmake/Modules/FindSYCLToolkit.cmake
Comment (min-jean-cho) [reply cmake/Modules/FindSYCLToolkit.cmake:55]: We may need to request compiler team to support `sycl.{dll, lib}` without version suffix to avoid hardcoding and maintaining particular version. In the meantime, I don't think there's an alternative?

File: cmake/Modules/FindMKLDNN.cmake
Comment (vpirogov) [reply cmake/Modules/FindMKLDNN.cmake:31]: `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT` [follows behavior](https://oneapi-src.github.io/oneDNN/dev_guide_build_options.html#host-compiler) defined by SYCL compiler itself. According to the [compiler documentation](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-1/fsycl-host-compiler.html#GUID-2151E257-78E0-44DC-8CB9-069F49869465) this means Intel DPC++ Compiler on Linux and Microsoft C/C++ Compiler on Windows.

+@densamoilov

Comment (min-jean-cho) [reply cmake/Modules/FindMKLDNN.cmake:31]: From offline discuss with @mdtoguchi, @vpirogov, the behavior of `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT` on Windows is:

- The DPC++ compiler's default host compiler is Clang on Windows.
- `-fsycl-host-compiler` is `OFF` by default on Windows.
- On Windows, with `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT`, `-fsycl-host-compiler` is [OFF](https://github.com/oneapi-src/oneDNN/blob/main/cmake/host_compiler.cmake#L23-L25) by default, meaning Clang is used for host compilation of oneDNN GPU library
- On Linux, with `ONEDNN_DPCPP_HOST_COMPILR=g++`, `-fsycl-host-compiler=g++` is [used](https://github.com/oneapi-src/oneDNN/blob/main/cmake/host_compiler.cmake#L103), meaning separate compilation with g++ as host compiler and DPC++ as device compiler.

The main concern of `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT` on Windows is ABI compatibility, if Clang compiled oneDNN GPU library is compatible to be linked with other MSVC compiled PyTorch libraries.

File: c10/util/Float8_fnuz_cvt.h
Comment (min-jean-cho) [reply c10/util/Float8_fnuz_cvt.h]: Thanks, per discuss with @fengyuan14 @EikanWang, updated for SYCL device code to use SYCL mapping `sycl::clz` of CUDA `__clz`.

File: cmake/Modules/FindMKLDNN.cmake
Comment (EikanWang) [reply cmake/Modules/FindMKLDNN.cmake:31]: It means that 3 different compilers will be involved: MSVC, Clang, and ICPX. Is it guaranteed that Clang is always compatible with MSVC for oneDNN scope?  @mdtoguchi, @vpirogov.

**Comment by min-jean-cho** [comment]
Bad rebase.

**Comment by lezcano** [comment]
too many people now tagged on this one after the bad rebase. Please open a new PR.

**Comment by min-jean-cho** [comment]
Hi @jgong5 @vpirogov @EikanWang @guangyey @dvrogozh @gujinghui @fengyuan14 , closing this PR due to bad rebase. Opened a new PR, https://github.com/pytorch/pytorch/pull/127390.

File: cmake/Modules/FindMKLDNN.cmake
Comment (mdtoguchi) [reply cmake/Modules/FindMKLDNN.cmake:31]: Generated binaries from `icx`, `icpx`, `icx-cl` on Windows will always be MSVC compatible unless there is an override via the `--target` option.  Use of `icx` or `icx-cl` is recommended if you require MSVC compatible command line options as well.  `icpx` will use Linux based options, but will target MSVC when run on Windows.

Comment (min-jean-cho) [reply cmake/Modules/FindMKLDNN.cmake:31]: Hi @mdtoguchi , let's move discussion to https://github.com/pytorch/pytorch/pull/127390. This PR had to be closed due to bad rebase.

**Pull request closed without merging**
